src/gui/sdlmain.cpp

```diff
@@ -1978,107 +1976,178 @@ static bool detect_resizable_window()
-		sdl.desktop.window.use_original_size = true;
-		return;
-	}
+static SDL_Point window_size_from_label(const std::string &pref, const SDL_Rect &desktop)
+{
+	int percent = 67; // default
+	if (starts_with("s", pref))
+		percent = 50;
+	else if (starts_with("m", pref) || pref == "default" || pref.empty())
+		percent = 67;
+	else if (starts_with("l", pref))
+		percent = 90;
+	else if (pref == "desktop")
+		percent = 100;
+	else
+		LOG_MSG("MAIN: The windowresolution '%s' is invalid, using 'default' instead",
+		        pref.c_str());
 
-	if (pref == "original") {
-		sdl.desktop.want_resizable_window = false;
-		sdl.desktop.window.use_original_size = true;
-		return;
-	}
+	const int w = ceil_sdivide(desktop.w * percent, 100);
+	const int h = ceil_sdivide(desktop.h * percent, 100);
+	return {w, h};
+}
 
-	int w = 0;
-	int h = 0;
-	// sscanf won't parse suffix after second integer
-	if (sscanf(pref.c_str(), "%dx%d", &w, &h) == 2) {
-		SDL_Rect bounds;
-		SDL_GetDisplayBounds(sdl.display_number, &bounds);
-		if (w > 0 && h > 0 && w <= bounds.w && h <= bounds.h) {
-			sdl.desktop.want_resizable_window = false;
-			sdl.desktop.window.use_original_size = false;
-			sdl.desktop.window.width = w;
-			sdl.desktop.window.height = h;
-			return;
-		} else {
-			LOG_MSG("MAIN: dimensions %dx%d are out of display "
-			        "bounds (%dx%d)",
-			        w, h, bounds.w, bounds.h);
-		}
-	}
+static SDL_Point window_size_from_conf(const char *windowresolution_val,
+                                       const SCALING_MODE scaling_mode,
+                                       const bool wants_43_aspect)
+{
+	assert(sdl.display_number >= 0);
+
+	// To-be-removed. For now, let DOS-side determine window-resolution
+	if (sdl.desktop.want_type == SCREEN_SURFACE)
+		return {0, 0};
+
+	SDL_Rect desktop;
+	SDL_GetDisplayBounds(sdl.display_number, &desktop);
+	assert(desktop.w >= 640 && desktop.h >= 480);
+
+	SDL_Point coarse_size;
+	const std::string pref = NormalizeConfValue(windowresolution_val);
+	if (pref.find('x') != std::string::npos)
+		coarse_size = window_size_from_resolution(pref, desktop);
+	else if (pref.find('%') != std::string::npos)
+		coarse_size = window_size_from_percent(pref, desktop);
+	else
+		coarse_size = window_size_from_label(pref, desktop);
+
+	sdl.desktop.want_resizable_window = detect_resizable_window();
 
-	LOG_MSG("MAIN: 'windowresolution = %s' is not a valid setting, using "
-	        "'original' instead", pref.c_str());
-	sdl.desktop.want_resizable_window = false;
-	sdl.desktop.window.use_original_size = true;
+	const auto refined_size = refine_window_size_with_mode(coarse_size,
+	                                                       scaling_mode,
+	                                                       wants_43_aspect);
+
+	LOG_MSG("MAIN: Selected window size %dx%d and %s aspect with %s constraints",
+	        refined_size.x, refined_size.y, wants_43_aspect ? "4:3" : "8:5",
```

Comment:
4:3 and 8:5 are magic numbers - perhaps move these to const-expr's too, and use those in the `refine_window_size_with_mode` function as well.